proof_test.go

```diff
@@ -188,3 +207,92 @@ func TestProof_MultipleLeaves(t *testing.T) {
 func safeAppend(id, data []byte) []byte {
 	return append(append(make([]byte, 0, len(id)+len(data)), id...), data...)
 }
+
+func TestVerifyLeafHashes_Err(t *testing.T) {
+	// create a sample tree
+	nmt := exampleTreeWithEightLeaves()
+	hasher := nmt.treeHasher
+	root, err := nmt.Root()
+	require.NoError(t, err)
+
+	// create an NMT proof
+	nID5 := namespace.ID{5, 5}
+	proof5, err := nmt.ProveNamespace(nID5)
+	require.NoError(t, err)
+	// corrupt the leafHash so that the proof verification fails during the root computation.
+	// note that the leaf at index 4 has the namespace ID of 5.
+	leafHash5 := nmt.leafHashes[4][:nmt.NamespaceSize()]
+
+	// create nmt proof for namespace ID 4
+	nID4 := namespace.ID{4, 4}
+	proof4, err := nmt.ProveNamespace(nID4)
+	require.NoError(t, err)
+	// corrupt the last node in the proof4.nodes, it resides on the right side of the proof4.end index.
+	// this test scenario makes the proof verification fail when constructing the tree root from the
+	// computed subtree root and the proof.nodes on the right side of the proof.end index.
+	proof4.nodes[2] = proof4.nodes[2][:nmt.NamespaceSize()-1]
+	leafHash4 := nmt.leafHashes[3]
+
+	tests := []struct {
+		name               string
+		proof              Proof
+		Hasher             *Hasher
+		verifyCompleteness bool
+		nID                namespace.ID
+		leafHashes         [][]byte
+		root               []byte
+		wantErr            bool
+	}{
+		{" wrong leafHash: not namespaced", proof5, hasher, true, nID5, [][]byte{leafHash5}, root, true},
+		{" wrong leafHash: incorrect namespace", proof5, hasher, true, nID5, [][]byte{{10, 10, 10, 10}}, root, true},
+		{" wrong proof.nodes: the last node has an incorrect format", proof4, hasher, false, nID4, [][]byte{leafHash4}, root, true},
+		//  the verifyCompleteness parameter in the verifyProof function should be set to false in order to bypass nodes correctness check during the completeness verification (otherwise it panics).
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			_, err := tt.proof.verifyLeafHashes(tt.Hasher, tt.verifyCompleteness, tt.nID, tt.leafHashes, tt.root)
+			assert.Equal(t, tt.wantErr, err != nil)
+		})
+	}
+}
+
+func TestVerifyInclusion_False(t *testing.T) {
+	// create a sample tree
+	nmt := exampleTreeWithEightLeaves()
+	hasher := nmt.treeHasher
+	root, err := nmt.Root()
+	require.NoError(t, err)
+
+	// create nmt proof for namespace ID 4
+	nID4 := namespace.ID{4, 4}
+	proof4, err := nmt.ProveNamespace(nID4)
+	require.NoError(t, err)
+	// proof4 is the inclusion proof for the leaf at index 3
+	leaf4WONamespace := nmt.leaves[3][nmt.NamespaceSize():] // the VerifyInclusion function expects the leaf without the namespace ID
```

Comment:
[question] what does the variable name here mean? It appears to truncate some leading bytes from the leaf so maybe leaf 4 without namespace? If so, proposal to rename 
```suggestion
	// proof4 is the inclusion proof for the leaf at index 3
	leaf4WithoutNamespace := nmt.leaves[3][nmt.NamespaceSize():] // the VerifyInclusion function expects the leaf without the namespace ID
```

[question] if VerifyInclusion expects the leaf without the namespace, why does this tampered test data fail?

Reply:
>[question] what does the variable name here mean? It appears to truncate some leading bytes from the leaf so maybe leaf 4 without namespace? If so, proposal to rename

In this variable name i.e., `leaf4WONamespace`, the `WO` part is the abbreviation of `without`, however,  it looks like unclear (since you asked), so I'm going to change it to the full version.

Reply:
>[question] what does the variable name here mean? It appears to truncate some leading bytes from the leaf so maybe leaf 4 without namespace? If so, proposal to rename

Please see the new variable name in https://github.com/celestiaorg/nmt/pull/136/commits/3f1b523e9d87d4254ced051aaf2f0e9f070f73e2